testing/matrix-sdk-integration-testing/src/tests/sliding_sync/room.rs

```diff
@@ -724,119 +724,85 @@ async fn test_delayed_decryption_latest_event() -> Result<()> {
+
     assert_eq!(bob_room.state(), RoomState::Joined);
+    assert!(bob_room.is_encrypted().await.unwrap());
 
+    // Get the room list of Alice.
     let alice_all_rooms = alice_sync_service.room_list_service().all_rooms().await.unwrap();
-    let (stream, entries) = alice_all_rooms
+    let (alice_room_list_stream, entries) = alice_all_rooms
         .entries_with_dynamic_adapters(10, alice.room_info_notable_update_receiver());
     entries.set_filter(Box::new(new_filter_all(vec![])));
-    pin_mut!(stream);
+    pin_mut!(alice_room_list_stream);
+
+    // The room list receives its initial rooms.
+    assert_let!(Some(diffs) = alice_room_list_stream.next().await);
+    assert_eq!(diffs.len(), 1);
+    assert_matches!(
+        &diffs[0],
+        VectorDiff::Reset { values: rooms } => {
+            assert_eq!(rooms.len(), 1);
+            assert_eq!(rooms[0].room_id(), room.room_id());
+        }
+    );
 
     // Send a message, but the keys won't arrive because to-device events are
     // stripped away from the server's response
     let event = bob_room.send(RoomMessageEventContent::text_plain("hello world")).await?;
 
-    // Wait shortly so the manual roominfo update is triggered before we load the
-    // stream.
-    sleep(Duration::from_secs(1)).await;
-
-    // Stream only has the initial Reset entry.
-    assert_let!(Some(diffs) = stream.next().await);
+    // Wait the message to be received by Alice.
+    assert_let!(Some(diffs) = alice_room_list_stream.next().await);
     assert_eq!(diffs.len(), 1);
     assert_matches!(
         &diffs[0],
-        VectorDiff::Reset { values: rooms } => {
-            assert_eq!(rooms.len(), 1);
-            assert_eq!(rooms[0].room_id(), alice_room.room_id());
+        VectorDiff::Set { index: 0, value: room } => {
+            // The latest event is not decrypted.
+            assert!(room.latest_event().await.is_none());
         }
     );
-    assert_pending!(stream);
-
-    // Latest event is not set yet
-    assert!(alice_room.latest_event().is_none());
 
     // Now we allow the key to come through
     *CUSTOM_RESPONDER.drop_todevice.lock().unwrap() = false;
 
-    // Wait for next sync
-    sleep(Duration::from_secs(3)).await;
-
-    // Latest event is set now
-    assert_eq!(alice_room.latest_event().unwrap().event_id(), Some(event.event_id));
-
-    // The stream has a single update
-    assert_let!(Some(diffs) = stream.next().await);
+    assert_let!(Some(diffs) = alice_room_list_stream.next().await);
     assert_eq!(diffs.len(), 1);
     assert_matches!(
         &diffs[0],
         VectorDiff::Set { index: 0, value: room } => {
-            assert_eq!(room.room_id(), alice_room.room_id());
+            // The latest event is now decrypted!
+            assert_eq!(room.latest_event().await.unwrap().event_id().unwrap(), event.event_id);
         }
     );
 
-    assert_pending!(stream);
+    assert_pending!(alice_room_list_stream);
```

Comment:
add a small wait before this assert_pending, just in case the final stream update is delayed.